src/Kernel.php

```diff
@@ -52,92 +50,80 @@
+            $state = loadState($message, $definition, $snapshotStore);
 
-            $events = loadEvents($definition, $aggregateId, $nextVersion, $eventStoreFactory);
+            $nextVersion = empty($state) ? 1 : $state[$definition->versionName()] + 1;
+
+            $events = loadEvents($eventStore, $definition, $aggregateId, $nextVersion);
 
             return $definition->reconstituteState($state, $events);
         };
 
-        $handleCommand = function (Message $message) use ($stateResolver, $commandMap): array {
+        $handleCommand = function (Message $message) use ($stateResolver, $commandMap): ImmList {
             $handler = getHandler($message, $commandMap);
 
             $events = $handler($stateResolver, $message);
 
             if (! is_array($events)) {
-                throw new RuntimeException('The handler did not return an array');
+                throw new \RuntimeException('The command handler did not return an array');
             }
 
-            return $events;
+            return ImmList(...$events);
         };
 
-        $persistEvents = function (array $events) use ($eventStoreFactory, $message, $getDefinition): array {
-            $definition = $getDefinition($message);
+        $enrichEvents = function (ImmList $events) use ($message, $definition, $aggregateId): Kind {
+            $enricher = getEnricherFor($definition, $aggregateId, $message);
 
-            return persistEvents($events, $eventStoreFactory, $definition, $definition->extractAggregateId($message));
+            return $events->map($enricher);
         };
 
-        return pipeline(
-            $handleCommand,
-            $persistEvents
-        )($message);
-    };
-}
+        $persistEvents = function (ImmList $events) use ($eventStore, $definition, $message, $aggregateId): Kind {
+            return persistEvents($events, $eventStore, $definition, $aggregateId);
+        };
 
-const pipeline = 'Prooph\Micro\Kernel\pipeline';
+        $publishEvents = function (ImmList $events) use ($publisher): Kind {
+            if ($events->isEmpty() || null === $publisher) {
+                return $events;
+            }
 
-function pipeline(callable $firstCallback, callable ...$callbacks): callable
-{
-    array_unshift($callbacks, $firstCallback);
+            return $events->map($publisher);
+        };
 
-    return function ($value = null) use ($callbacks) {
-        try {
-            $result = array_reduce($callbacks, function ($accumulator, callable $callback) {
-                return $callback($accumulator);
-            }, $value);
-        } catch (Throwable $e) {
-            return $e;
-        }
+        $pipe = function () use ($message, $handleCommand, $enrichEvents, $persistEvents, $publishEvents) {
+            return Function1($handleCommand)
+                ->andThen($enrichEvents)
+                ->andThen($persistEvents)
+                ->andThen($publishEvents)
+                ->run($message);
```

Comment:
cool! 😄 
This can also be written with `compose`, thus:
```php
return compose(
    $handleCommand,
    $enrichEvents,
    $persistEvents,
    $publishEvents)($message)
```